source/simulator/core.cc

```diff
@@ -1083,28 +1094,11 @@ namespace aspect
             }
         }
       coupling[x.temperature][x.temperature] = DoFTools::always;
-      for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
-        {
-          const AdvectionField adv_field (AdvectionField::composition(c));
-          const typename Parameters<dim>::AdvectionFieldMethod::Kind method = adv_field.advection_method(introspection);
-          switch (method)
-            {
-              case Parameters<dim>::AdvectionFieldMethod::fem_field:
-                coupling[x.compositional_fields[c]][x.compositional_fields[c]] = DoFTools::always;
-                break;
-
-              case Parameters<dim>::AdvectionFieldMethod::particles:
-                coupling[x.compositional_fields[c]][x.compositional_fields[c]] = DoFTools::none;
-                break;
-
-              case Parameters<dim>::AdvectionFieldMethod::static_field:
-                coupling[x.compositional_fields[c]][x.compositional_fields[c]] = DoFTools::none;
-                break;
 
-              default:
-                Assert(false,ExcNotImplemented());
-            }
-        }
+      // if we have at least one compositional field that is a FEM field, we create a matrix block in the first
+      // compositional block
+      if (have_fem_compositional_field)
+        coupling[x.compositional_fields[0]][x.compositional_fields[0]] = DoFTools::always;
```

Comment:
can you add to the comment that the sparsity patterns for the remaining compositional (diagonal) blocks will remain empty this way, but that the corresponding matrix blocks will then just re-direct to the one block created here? otherwise, the purpose of this piece of code will remain a bit obscure and require digging up code elsewhere.